x/ibc/20-transfer/kepper/keeper.go

```diff
@@ -0,0 +1,166 @@
+func (k Keeper) ReceiveTransfer(ctx sdk.Context, data types.TransferPacketData, destPort, destChannel, srcPort, srcChannel string) sdk.Error {
+
+	receiverAddr, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return sdk.NewError(sdk.CodespaceType(types.DefaultCodespace), types.CodeInvalidReceiver, "invalid receiver address")
+	}
+
+	if data.Source {
+		// mint tokens
+
+		// check the denom prefix
+		if !strings.HasPrefix(data.Denomination, getDenomPrefix(destPort, destChannel)) {
+			return sdk.NewError(sdk.CodespaceType(types.DefaultCodespace), types.CodeIncorrectDenom, "incorrect denomination")
+		}
+
+		_, err := k.bk.AddCoins(ctx, receiverAddr, sdk.Coins{sdk.NewCoin(data.Denomination, data.Amount)})
+		if err != nil {
+			return err
+		}
+
+	} else {
+		// unescrow tokens
+
+		// check the denom prefix
+		prefix := getDenomPrefix(srcPort, srcChannel)
+		if !strings.HasPrefix(data.Denomination, prefix) {
+			return sdk.NewError(sdk.CodespaceType(types.DefaultCodespace), types.CodeIncorrectDenom, "incorrect denomination")
+		}
+
+		escrowAddress := getEscrowAddress(destChannel)
+		err := k.bk.SendCoins(ctx, escrowAddress, receiverAddr, sdk.Coins{sdk.NewCoin(data.Denomination[len(prefix):], data.Amount)})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (k Keeper) createOutgoingPacket(ctx sdk.Context, seq uint64, srcPort, srcChan, destPort, destChan string, denom string, amount sdk.Int, sender sdk.AccAddress, receiver string, source bool) sdk.Error {
+	if source {
+		// escrow tokens
+
+		// get escrow address
+		escrowAddress := getEscrowAddress(srcChan)
+
+		// check the denom prefix
+		prefix := getDenomPrefix(destPort, destChan)
+		if !strings.HasPrefix(denom, prefix) {
+			return sdk.NewError(sdk.CodespaceType(types.DefaultCodespace), types.CodeIncorrectDenom, "incorrect denomination")
+		}
+
+		err := k.bk.SendCoins(ctx, sender, escrowAddress, sdk.Coins{sdk.NewCoin(denom[len(prefix):], amount)})
+		if err != nil {
+			return err
+		}
+
+	} else {
+		// burn vouchers from sender
+
+		// check the denom prefix
+		prefix := getDenomPrefix(srcPort, srcChan)
+		if !strings.HasPrefix(denom, prefix) {
+			return sdk.NewError(sdk.CodespaceType(types.DefaultCodespace), types.CodeIncorrectDenom, "incorrect denomination")
+		}
+
+		_, err := k.bk.SubtractCoins(ctx, sender, sdk.Coins{sdk.NewCoin(denom, amount)})
+		if err != nil {
+			return err
+		}
+	}
+
+	// build packet
+	packetData := types.TransferPacketData{
+		Denomination: denom,
+		Amount:       amount,
+		Sender:       sender.String(),
+		Receiver:     receiver,
+		Source:       source,
+	}
+
+	packetDataBz, err := packetData.MarshalJSON()
+	if err != nil {
+		return sdk.NewError(sdk.CodespaceType(types.DefaultCodespace), types.CodeInvalidPacketData, "invalid packet data")
+	}
+
+	packet := channeltypes.NewPacket(seq, uint64(ctx.BlockHeight())+DefaultPacketTimeout, srcPort, srcChan, destPort, destChan, packetDataBz)
```

Comment:
how ctx.BlockHeight() useful here ?

Reply:
yeah, the timeout block height should be the destination chain block height.

Reply:
> how ctx.BlockHeight() useful here ?, if destination chain blockheight > (ctx.blockheight+DefaultPacketTimeout)

How about using the latest consensus height of the counterparty?